signaltometricsconnector: support gauges

simitt opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Add support for parsing `gauges` from any signal type with last value aggregation per batch. 
For now only numerical values are supported. Parsing timestamps is not yet supported. 


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Relates to #37093

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Automated unit tests have been added. 

<!--Describe the documentation added.-->
#### Documentation
README was updated
<!--Please delete paragraphs that you did not use before submitting.-->

simitt committed:
signaltometricsconnector: support gauges

Add support for gauges with last value aggregation.
relates to #37093

lahsivjar commented on connector/signaltometricsconnector/README.md:
```diff
@@ -91,6 +92,31 @@ sum:
 
+#### Gauge
+
+Gauge metrics are supported for logs and allow extracting numeric values from log bodies using Grok patterns via the `ExtractGrokPatterns` OTTL function. The gauges are aggregated by last value. 
```
Are we supporting only logs or all signal types? I was thinking that we would add support for gauges for all signal types.

simitt replied:
I only added support for logs, will look at other signal types also.

lahsivjar replied:
The aggregator instance is shared and called for all signal types, so, I think the code should already be working for all signal types.

lahsivjar commented on connector/signaltometricsconnector/README.md:
````diff
@@ -91,6 +92,31 @@ sum:
+
+```yaml
+gauge:
+  value: ExtractGrokPatterns(target, pattern)
````
What do you think about keeping the value pattern only support numbers (`int64` and `float64`)? I think the complexity of handling a map as return is not worth it and users can easily do something like `ExtractGrokPatterns(target, pattern)["key"]` to convert it to a number I think.

lahsivjar commented on connector/signaltometricsconnector/config/config.go:
```diff
@@ -299,6 +319,36 @@ func validateMetricInfo[K any](mi MetricInfo, parser ottl.Parser[K]) error {
+		if _, err := parser.ParseValueExpression(mi.Gauge.Value); err != nil {
+			return fmt.Errorf("failed to parse value OTTL expression for gauge: %w", err)
+		}
+		// Custom validation: fail if ExtractGrokPatterns is used with multiple patterns or configured with unsupported type
+		if strings.Contains(mi.Gauge.Value, "ExtractGrokPatterns") {
```
I am thinking to limit the support for defined OTTL expressions to return only a numerical value - anything else could be rejected during runtime with an error log (see the above comment for more details).

lahsivjar reviewed (commented):
Added a few discussion points on the supported return types for the `value` expression.

lahsivjar commented on connector/signaltometricsconnector/internal/aggregator/aggregator.go:
```diff
@@ -87,6 +89,28 @@ func (a *Aggregator[K]) Aggregate(
+		if err != nil {
+			return fmt.Errorf("failed to execute OTTL value for gauge: %w", err)
+		}
+		v, ok := raw.(pcommon.Map)
```
To support other use cases for gauges, IMO, we should definitely support numerical types.

lahsivjar commented on connector/signaltometricsconnector/internal/aggregator/aggregator.go:
```diff
@@ -145,11 +169,27 @@ func (a *Aggregator[K]) Finalize(mds []model.MetricDef[K]) {
 				dp.Copy(a.timestamp, destCounter.DataPoints().AppendEmpty())
 			}
 		}
+		if md.Gauge == nil {
+			continue
+		}
```
We should probably put this inside the for loop with `break`, otherwise the `delete` bit will not be invoked for anything other than gauge.

simitt committed:
Address review comments:

- allow for all signals
- only support numbers in OTTL expressions rather than Grok maps.
- move check into loop

simitt commented:
@lahsivjar thanks for the initial round of feedback, moving this out of draft now after addressing your feedback.

lahsivjar commented on connector/signaltometricsconnector/README.md:
```diff
@@ -66,9 +66,10 @@ signaltometrics:
+component can produce the following metric types for each signal type:
 
 - [Sum](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums)
+- [Gauge](#gauge)
```
```suggestion
- [Gauge](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#gauge)
```

For consistency as we seem to be linking the data-model docs here.

lahsivjar commented on connector/signaltometricsconnector/README.md:
```diff
@@ -91,6 +92,43 @@ sum:
+  the signal. Only OTTL expressions that return a value are accepted. The returned 
+  value determines the value type of the `gauge` metric (`int` or `double`).
+  - For logs: Use e.g. `ExtractGrokPatterns` with a single key selector (see below). 
+  - For other signals: Use a field such as `value_int`, `value_double`, or a valid OTTL expression for your context.
```
```suggestion
  - For other signals: Use a field such as `value_int`, `value_double`, or a valid OTTL expression.
```

lahsivjar commented on connector/signaltometricsconnector/config/config.go:
```diff
@@ -299,6 +318,19 @@ func validateMetricInfo[K any](mi MetricInfo, parser ottl.Parser[K]) error {
+			return fmt.Errorf("failed to parse value OTTL expression for gauge: %w", err)
+		}
+		// if ExtractGrokPatterns is used, validate the key selector
+		if strings.Contains(mi.Gauge.Value, "ExtractGrokPatterns") {
+			// Ensure a [key] selector is present after ExtractGrokPatterns
+			if !regexp.MustCompile(`ExtractGrokPatterns\([^)]*\)\s*\[[^\]]+\]`).MatchString(mi.Gauge.Value) {
+				return fmt.Errorf("ExtractGrokPatterns: a single key selector[key] is required for signal to gauge")
+			}
+		}
```
This check would be true for all configured metric types so maybe in a follow-up PR we can generalize this as a common validation for `ExtractGrokPatterns`.

lahsivjar replied:
This is not that important though and we can just create an issue and leave it for future too as we might need to think about how to (or if to) handle config validation for all funcs that return map (or other non-primitives)

simitt replied:
created https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/40118

lahsivjar reviewed (approved):
LGTM, a few minor README comments! Thanks for addressing all the review points.

simitt committed:
Apply suggestions from code review

Co-authored-by: Vishal Raj <vishal.raj@elastic.co>

lahsivjar commented:
@simitt There are some minor lint errors and it also needs a changelog. More details on adding changelog is available here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c37a07feb5a5469691489e3cd37097c059bbd60d/CONTRIBUTING.md?plain=1#L72-L76

simitt committed:
Add changelog entry

simitt committed:
Fix linting issues

lahsivjar commented on .chloggen/signaltometrics-logs-to-gauge.yaml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
```
**nit**: Can we change the name of the changelog file to `signaltometrics-gauge-support.yaml`? (by default the name of the file is taken from the branch, however, AFAIK, it can be changed)

simitt replied:
Updated!

simitt committed:
rename changelog

ChrsMark commented on connector/signaltometricsconnector/config/config.go:
```diff
@@ -299,6 +318,19 @@ func validateMetricInfo[K any](mi MetricInfo, parser ottl.Parser[K]) error {
+		if _, err := parser.ParseValueExpression(mi.Gauge.Value); err != nil {
+			return fmt.Errorf("failed to parse value OTTL expression for gauge: %w", err)
+		}
+		// if ExtractGrokPatterns is used, validate the key selector
```
Why only this specific ottl function is checked when any kind of OTTL function is allowed? Same check should be applied for [extractpatterns](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md#extractpatterns) function, right?

I wonder if validation makes sense here since any kind of OTTL function and I'm not sure if we know before hand which of those should need validation or not.

lahsivjar replied:
I am also on the fence about the need to have such validation, we had a similar discussion [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40113#discussion_r2093058092) which resulted in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/40118

ChrsMark commented on connector/signaltometricsconnector/config/config.go:
```diff
@@ -299,6 +318,19 @@ func validateMetricInfo[K any](mi MetricInfo, parser ottl.Parser[K]) error {
+		// if ExtractGrokPatterns is used, validate the key selector
+		if strings.Contains(mi.Gauge.Value, "ExtractGrokPatterns") {
+			// Ensure a [key] selector is present after ExtractGrokPatterns
+			if !regexp.MustCompile(`ExtractGrokPatterns\([^)]*\)\s*\[[^\]]+\]`).MatchString(mi.Gauge.Value) {
```
nit: even if it's only executed once maybe we can consider moving the `MustCompile` part to var at package level to be more idiomatic (no performance impact since config validation is only executed once).

simitt committed:
move regex into var

simitt committed:
Add ExtractPatterns to config test

ChrsMark committed:
Merge branch 'main' into signaltometrics-logs-to-gauge
